src/julia-syntax.scm

```diff
@@ -4064,9 +4085,11 @@ f(x) = yt(x)
                         (type-name  (or (get namemap name #f)
                                         (and name
                                              (symbol (string (if (= (string.char (string name) 0) #\#)
-                                                                 ""
-                                                                 "#")
-                                                             name "#" (current-julia-module-counter))))))
+                                                                  (if (anon-function-name? (string name))
+                                                                    (string "#" (current-julia-module-counter parsed-method-stack))
+                                                                    name)
+                                                                  (string "#" name))
+                                                              "#" (current-julia-module-counter parsed-method-stack))))))
```

Comment:
If I'm parsing this right, it looks like this might call `current-julia-module-counter` twice for a single identifier? I don't see why that would be necessary.

Reply:
Types of anonymous functions are currently printed as `#2#3` on master.

With this PR, they get printed as `#foo##2#foo##3` (assuming the anonymous function is enclosed by a top-level function `foo`).

I'm calling it twice so that:

- We ensure the pattern `<something_1>#<something_2>` for callable type names doesn't change.
- We ensure the string `something_1` gets "canonicalized", in the sense of getting an integer suffix based on the table that's per top-level function name (instead of an integer suffix that's per module).

If we can relax the constraint of keeping the callable type names as `<something_1>#<something_2>`, then we can probably just call `current-julia-module-counter` once to generate a type name like `#foo##2`, instead of `#foo##2#foo##3`.

Reply:
@topolarity and @NHDaly didn't consider this duplication to be a merge blocker. In any case, opened an issue so that we can track this in the future: https://github.com/JuliaLang/julia/issues/54230.